ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/State.hs

```diff
@@ -0,0 +1,385 @@
+  where
+    -- TODO Is uninterruptibleMask_ absolutely necessary here?
+    getSt :: m (OpenState blockId h)
+    getSt = uninterruptibleMask_ $ takeMVar varInternalState >>= \case
+      DbOpen ost -> return ost
+      DbClosed   -> do
+        putMVar varInternalState DbClosed
+        throwM $ UserError ClosedDBError
+
+    putSt :: OpenState blockId h -> ExitCase (OpenState blockId h) -> m ()
+    putSt ost ec = do
+        -- It is crucial to replace the MVar.
+        putMVar varInternalState st'
+        unless (dbIsOpen st') $ closeOpenHandles hasFS ost
+      where
+        st' = case ec of
+          ExitCaseSuccess ost' -> DbOpen ost'
+
+          -- When something goes wrong, close the VolatileDB for safety.
+          -- Except for user errors, because they stem from incorrect use of
+          -- the VolatileDB.
+          --
+          -- NOTE: we only modify the VolatileDB in a background thread of
+          -- the ChainDB, not in per-connection threads that could be killed
+          -- at any point. When an exception is encountered while modifying
+          -- the VolatileDB in the background thread, or that background
+          -- thread itself is killed with an async exception, we will shut
+          -- down the node anway, so it is safe to close the VolatileDB here.
+          ExitCaseAbort        -> DbClosed
+          ExitCaseException ex
+            | Just (UserError {}) <- fromException ex
+            -> DbOpen ost
+            | otherwise
+            -> DbClosed
+
+-- | Perform an action that accesses the internal state of an open database.
+--
+-- In case the database is closed, a 'ClosedDBError' is thrown.
+--
+-- In case an 'UnexpectedError' is thrown while the action is being run, the
+-- database is closed to prevent further appending to a database in a
+-- potentially inconsistent state. All other exceptions will leave the
+-- database open.
+withOpenState
+  :: forall blockId m r. (HasCallStack, IOLike m)
+  => VolatileDBEnv m blockId
+  -> (forall h. HasFS m h -> OpenState blockId h -> m r)
+  -> m r
+withOpenState VolatileDBEnv {hasFS = hasFS :: HasFS m h, varInternalState} action = do
+    (mr, ()) <- generalBracket open (const close) (tryVolDB . access)
+    case mr of
+      Left  e -> throwM e
+      Right r -> return r
+  where
+    open :: m (OpenState blockId h)
+    open = readMVar varInternalState >>= \case
+      DbOpen ost -> return ost
+      DbClosed   -> throwM $ UserError ClosedDBError
+
+    -- close doesn't take the state that @open@ returned, because the state
+    -- may have been updated by someone else since we got it (remember we're
+    -- using 'readMVar' here, 'takeMVar'). So we need to get the most recent
+    -- state anyway.
+    close :: ExitCase (Either VolatileDBError r)
+          -> m ()
+    close ec = case ec of
+      ExitCaseAbort                               -> return ()
+      ExitCaseException _ex                       -> return ()
```

Comment:
At `modifyOpenState` we close the db in this case. Why is it different here?

Reply:
In `modifyOpenState`, we might have written something to disk, so we are no longer sure that our in-memory state matches our on-disk state. We close as a safety measure. Note that we expect `ExitCaseException` mainly to be triggered by async exceptions, in which case we should shut down the database anyway. This is related to #1681.

In `withOpenState`, we are not modifying the state, so an async exception must not and *should* not close the database, as the per-connection threads reading from the VolatileDB *can* be killed.